internal/operator-controller/authorization/rbac.go

```diff
@@ -0,0 +1,631 @@
+		return nil
+	}
+
+	roleRef := rbacinternal.RoleRef{}
+	err := rbacv1helpers.Convert_v1_RoleRef_To_rbac_RoleRef(&roleBinding.RoleRef, &roleRef, nil)
+	if err != nil {
+		return err
+	}
+	if rbacregistry.BindingAuthorized(ctx, roleRef, roleBinding.Namespace, ec.authorizer) {
+		return nil
+	}
+
+	rules, err := ec.ruleResolver.GetRoleReferenceRules(ctx, roleBinding.RoleRef, roleBinding.Namespace)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	switch roleRef.Kind {
+	case "ClusterRole":
+		if manifestClusterRole, ok := ec.extraClusterRoles[types.NamespacedName{Name: roleBinding.RoleRef.Name}]; ok {
+			rules = append(rules, manifestClusterRole.Rules...)
+		}
+	case "Role":
+		if manifestRole, ok := ec.extraRoles[types.NamespacedName{Namespace: roleBinding.Namespace, Name: roleBinding.RoleRef.Name}]; ok {
+			rules = append(rules, manifestRole.Rules...)
+		}
+	}
+
+	if err := validation.ConfirmNoEscalation(ctx, ec.ruleResolver, rules); err != nil {
+		return err
+	}
+	return nil
+}
+
+var fullAuthority = []rbacv1.PolicyRule{
+	{Verbs: []string{"*"}, APIGroups: []string{"*"}, Resources: []string{"*"}},
+	{Verbs: []string{"*"}, NonResourceURLs: []string{"*"}},
+}
+
+// TODO: Investigate replacing this regex parsing with structured error handling once there are
+//
+//	structured RBAC errors introduced by https://github.com/kubernetes/kubernetes/pull/130955.
+//
+// parseEscalationErrorForMissingRules attempts to extract specific RBAC permissions
+// that were denied due to escalation prevention from a given error's text.
+// It returns the list of extracted PolicyRules and an error detailing the escalation attempt
+// and any resolution errors found.
+// Note: If parsing is successful, the returned error is derived from the *input* error's
+// message, not an error encountered during the parsing process itself. If parsing fails due to an unexpected
+// error format, a distinct parsing error is returned.
+func parseEscalationErrorForMissingRules(ecError error) ([]rbacv1.PolicyRule, error) {
+	// errRegex captures the missing permissions and optionally resolution errors from an escalation error message
+	// Group 1: The list of missing permissions
+	// Group 2: Optional resolution errors
+	errRegex := regexp.MustCompile(`(?s)^user ".*" \(groups=.*\) is attempting to grant RBAC permissions not currently held: (.*)(?:; resolution errors: (.*))?$`)
+	// permRegex extracts the details (APIGroups, Resources, Verbs) of individual permissions listed within the error message
+	permRegex := regexp.MustCompile(`{APIGroups:\[([^\]]*)\], Resources:\[([^\]]*)\], Verbs:\[([^\]]*)\]}`)
+
+	errString := ecError.Error()
+	errMatches := errRegex.FindStringSubmatch(errString) // Use FindStringSubmatch for single match expected
+
+	// Check if the main error message pattern was matched and captured the required groups
+	// We expect at least 3 elements: full match, missing permissions, resolution errors (can be empty)
+	if len(errMatches) < 3 {
```

Comment:
This should explicitly check for 2 or 3 only.

Reply:
done in a167ef494634aace8dfa3cbde63949eb88867ced